File: go/pkg/utils/seeder.go
package utils

import (
	"context"
	"fmt"
	"sync"

	"github.com/openfga/go-sdk/client"
)

func WriteBatches(ctx context.Context, fgaClient *client.OpenFgaClient, tuples []client.ClientTupleKey, batchSize, concurrency int) error {
	var batches [][]client.ClientTupleKey
	for i := 0; i < len(tuples); i += batchSize {
		end := i + batchSize
		if end > len(tuples) {
			end = len(tuples)
		}
		batches = append(batches, tuples[i:end])
	}

	queue := make(chan []client.ClientTupleKey, len(batches))
	for _, batch := range batches {
		queue <- batch
	}
	close(queue)

	var wg sync.WaitGroup
	errChan := make(chan error, concurrency)

	for i := 0; i < concurrency; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for batch := range queue {
				body := client.ClientWriteRequest{
					Writes: batch,
				}
				_, err := fgaClient.Write(ctx).Body(body).Execute()
				if err != nil {
					errChan <- fmt.Errorf("error seeding batch: %w", err)
				}
			}
		}()
	}

	wg.Wait()
	close(errChan)

	for err := range errChan {
		if err != nil {
			return err
		}
	}

	return nil
}

File: src/utils/seeder.js
const { BATCH_SIZE, CONCURRENCY } = require('../config');

const writeBatches = async (fgaClient, allTuples) => {
  const queue = [];
  for (let i = 0; i < allTuples.length; i += BATCH_SIZE) {
    queue.push(allTuples.slice(i, i + BATCH_SIZE));
  }

  async function worker() {
    while (queue.length > 0) {
      const batch = queue.shift();
      try {
        await fgaClient.writeTuples(batch);
      } catch (e) {
        console.error('Error seeding batch:', e);
      }
    }
  }

  const workers = [];
  for (let i = 0; i < CONCURRENCY; i++) workers.push(worker());
  await Promise.all(workers);
};

module.exports = { writeBatches };